onnx/defs/shape_inference.h

```diff
@@ -75,11 +78,23 @@ class InferenceError final : public std::runtime_error {
   std::string expanded_message_;
 };
 
+class ShapeInferenceError final : public InferenceError {
```

Comment:
Re. the naming: I wonder if it is better to call these `ShapeError` and `TypeError`. (I realize that this naming comes from the original macro names.)

Also, I think these might not be enough: some errors (like missing attributes or wrong attribute-value) are not really "shape error". So, I wonder if we should either rename `ShapeError` to `ValueError` or at least add another exception called `ValueError` or something more generic.